c10/core/impl/cow/state_machine.cpp

```diff
@@ -0,0 +1,240 @@
+#include <c10/core/impl/cow/state_machine.h>
+
+#include <c10/util/Exception.h>
+#include <c10/util/Optional.h>
+
+#include <mutex>
+
+namespace c10::impl {
+
+class cow::StateMachine::Impl {
+ public:
+  /** @see cow::StateMachine::maybe_bump */
+  auto maybe_bump(cow::ShadowStorage* maybe_shadow_storage) -> void;
+
+  /** @see cow::StateMachine::simulate_lazy_copy */
+  auto simulate_lazy_copy(cow::ShadowStorage* maybe_shadow_storage)
+      -> intrusive_ptr<cow::ShadowStorage>;
+
+ private:
+  friend class cow::StateMachine;
+
+  // Guards all the state.
+  std::mutex mtx_;
+  // How many writes have been applied to the storage.
+  cow::StateMachine::Generation physical_generation_ = 0;
```

Comment:
There is also a generation on the ShadowStorageImpl as well, so it would be helpful to point out what exactly the distinction between this and the generation stored on the default shadow storage is

Reply:
Please take a another look.